### ISSUE TITLE
breaks npm install -g npm in node 0.8

### ISSUE THREAD

**Issue by calvinmetcalf** [npm/npm#9668]
previously in readable-stream we test all the way back to node 0.8, previoulsy we could still use carrots by doing npm install -g npm, this worked great until 3.3.3 which fails with a `Error: Cannot find module 'are-we-there-yet'` see https://travis-ci.org/nodejs/readable-stream/jobs/81113430

**Comment by othiym23** [comment]
So there's a dilemma here. The problem isn't with carets, it's that there's something about how `npm@3` packages itself and its bundled dependencies that `npm@1.2` just can't deal with, probably related to the flattened dependency tree. There were a bunch of bugs in dealing with packaged bundled dependencies well up through `npm@2.10`, so I'm unsurprised to hear that 1.2 has problems. I'm not sure it's worth the effort to figure out what's triggering the problem, because aside from Travis, people are going to be doing single-step upgrades from `npm@1.2.30` to `npm@3` very, very rarely, and because there's a simple workaround (read on). That said, you're the third or fourth person to run into this today (@isaacs also got tripped up on this, which is bad for obvious reasons), so I imagine a lot of people who care about legacy testing will be affected by this. It would be great if somebody did a little digging and figured out what the problem is here so we could see how hard it's going to be to fix (which, obviously, can't be done by patching the version of npm bundled with old Nodes).

The workaround is simple enough: you'll need to add another step to your `.travis.yml` to have it first upgrade from the npm with 0.8 to `npm@2`, and from there upgrade again to `npm@latest`. Or, if you want, from the bundled version to `npm@1.4.28`. Neither have any problem upgrading to `npm@3`.

**Comment by leetrout** [comment]
> people are going to be doing single-step upgrades from npm@1.2.30 to npm@3 very, very rarely

FWIW just hit this bringing a vanilla Ubuntu 14.04 box online in Vagrant.

**Comment by othiym23** [comment]
@leetrout The above discussion aside, you're going to have a _much_ better time if you ditch the Ubuntu-packaged Node and npm and go with the [NodeSource distribution](https://github.com/nodesource/distributions).

**Comment by calvinmetcalf** [comment]
this did it dd40a2954b9110fce9909b571db30a3d24a31111

**Comment by krotscheck** [comment]
Unfortunately the nodesource distribution is not an option for those of us who, due to security policy, are locked to LTS releases and upstream packages. Which pretty much means any financial institution, and those who serve them.

**Comment by krotscheck** [comment]
I suspect this problem will "go away" once debian and/or centos and/or suse adopt the Node 4 LTS branch, but unless the linux/node foundation get that out in time for distro lockdown, that could be a year or two.

**Comment by othiym23** [comment]
@krotscheck It's entirely likely that Node 4 LTS will only include `npm@2`, and if you're restricted to using only packaged version of utilities like npm, this issue will not affect you. Either way, the workaround listed above will fix this, and there's a possibility that we'll figure out how to package npm in such a way that very old versions like `npm@1.2` and `npm@1.3` will be able to install it.

**Comment by krotscheck** [comment]
@othiym23 Actually, the workaround doesn't quite work. Take a look:

```
  $ sudo su -
  $ apt-get install npm
  $ npm --version
  1.3.10
  $ which npm
  /usr/bin/npm
  $ npm -g prefix
  /usr

  $ npm install -g --upgrade npm@2
  $ npm --version
  2.14.5
  $ which npm
  /usr/local/bin/npm
  $ npm -g prefix
  /usr/local

  $ npm install -g --upgrade npm
  $ npm --version
  2.14.5
  $ which npm
  /usr/local/bin/npm
  $ npm -g prefix
  /usr
  $ /usr/bin/npm --version
  3.3.3
```

More detailed output here -> http://paste.openstack.org/show/473108/ - it appears that NPM's default prefix is different between versions of npm.

**Comment by fungi** [comment]
Are you sure it's not simply that the debian package overrides the default prefix, and what you're seeing is switching between the debian default behavior in 1.x and the upstream NPM default behavior when running the 2.x you installed with it?

**Comment by fungi** [comment]
Just confirmed, Debian's package ships a custom /usr/share/npm/npmrc file containing prefix=/usr/local to achieve the initial behavior you observed.

**Comment by othiym23** [comment]
@krotscheck The problem is that the Debian-packaged version of npm has a whole bunch of custom tweaks that aren't included in the version supported by npm, which get removed the first time you upgrade, so different things happen on the second run of `npm install` (also, `--upgrade` isn't a configuration option used by npm, for future reference). If you figure out how to do that in fewer steps than it took you this time, write it up as a script or a gist and draw our attention to it (or just add it to the [Troubleshooting guide](https://github.com/npm/npm/wiki/Troubleshooting) yourself).

**Comment by krotscheck** [comment]
@othiym23 Ok, so, who at NPM is responsible for ensuring that the distro packages are sanely configured? This seems like something y'all would care about, no?

**Comment by othiym23** [comment]
That would probably be me, for now, and yes, we do care about how npm is packaged, but we haven't historically had much control over the process. Both Debian and Fedora have fairly elaborate policies and standards for packaging, and one of the big issues is that those policies weren't built with Node in mind. Node and npm both move fast by the standards of distro maintainers, and while I've tried to persuade various maintainers that npm is one big package instead of many small packages, I haven't had much luck. I'd love to find a way to work more directly with npm's downstream maintainers, but also the amount of time I can budget for that project is small, which seems to be true from their side as well.

**Comment by iarna** [comment]
To the original issue, the root cause goes thusly:

1. npm@3 uses itself to install its modules, resulting in a flat tree.
2. npm bundles its dependencies, and the bundledDependencies in the package.json are only the things npm itself depends on directly.
3. npm@3 has special bundling magic, so that when it `npm pack`s (done internally on publish) it knows that it shoiuld include the deps of the bundled deps.
4. npm@1.2 of course lacks this knowledge of what's bundled. That wouldn't be a problem (it isn't for later versions of npm) but it still fails because of a quirk of 1.2's install process, which is:
  1. Download a package from somewhere… registry, git, remote tarball…
  2. Extract the package into a temporary location
  3. Run `npm pack` in this location and add the resulting tarball to the cache.
  4. This last step, of course fails, because npm@1 can't pack packages that have bundled and flattened deps.

**Comment by moneypenny** [comment]
Just hit this myself bringing up a fresh vagrant VM.

**Comment by othiym23** [comment]
We're going to address this issue by using `npm@2` to install `npm@>=3`'s dependencies, so that old versions of npm are able to resume doing a single-step upgrade to `npm@3`. npm will probably continue to do that for as long as the team continues to support Node 0.8 (because the versions of npm packaged with 0.10 don't have this problem, at least for sufficiently recent versions of 0.10). I don't have an exact date of when we're going to do that work, but it's now on the team's list of things to do.

I don't have a fixed idea of when npm will give up on Node 0.8, but it will probably coincide with some of the larger installations of Node 0.8 deciding (finally) to upgrade to a newer version of Node.
